### PR TITLE
XIVE test cases

### PR DESCRIPTION
**Pull request by kalshett**
Added new test cases for XIVE feature

Signed-off-by: Kalpana Shetty <kalshett@in.ibm.com>

### CODE REVIEW

**Comment by kalshett** [comment]
Test Result:

case 1: P10 in default mode(xive)
Result: 1-PASS, 2-PASS

[root@ltcden14-lp10 cpu]# avocado run xive.py
JOB ID     : cad9cbc800e02846fc0982e8f3cbf1b98ddddcf8
JOB LOG    : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T03.59-cad9cbc/job.log
 (1/2) xive.py:XIVE.test_intr_mode: PASS (0.48 s)
 (2/2) xive.py:XIVE.test_storeEOI: PASS (0.45 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T03.59-cad9cbc/results.html
JOB TIME   : 1.56 s

case 2: P10 in P9compat mode(xive)
Result: 1-PASS, 2-CANCEL

[root@ltcden14-lp10 cpu]# avocado run xive.py
JOB ID     : fec9537815143d3b187d37078a292366393e437d
JOB LOG    : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T03.10-fec9537/job.log
 (1/2) xive.py:XIVE.test_intr_mode: PASS (0.63 s)
 (2/2) xive.py:XIVE.test_storeEOI: CANCEL: storeEOI feature 'S' flag is not set for POWER9 / XIVE (0.60 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 1
JOB HTML   : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T03.10-fec9537/results.html
JOB TIME   : 1.82 s

case 3: P10 in P9compat_base mode(xics)
Result: 1-PASS, 2-CANCEL

[root@ltcden14-lp10 cpu]# avocado run xive.py
JOB ID     : 446abd64510ddb027b21c7fbde6f28c6203519ef
JOB LOG    : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T02.46-446abd6/job.log
 (1/2) xive.py:XIVE.test_intr_mode: PASS (0.59 s)
 (2/2) xive.py:XIVE.test_storeEOI: CANCEL: storeEOI feature is not Available for POWER9 / XICS (0.56 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 1
JOB HTML   : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T02.46-446abd6/results.html
JOB TIME   : 1.80 s


case 4: P9 in default mode(xics)
Result: 1-PASS, 2-CANCEL

/home/kal/avocado-fvt-wrapper/tests/avocado-misc-tests/cpu # avocado run xive.py
JOB ID     : 2ec4c223b0b4e6229513b4bd57c76b3116952fd9
JOB LOG    : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T03.58-2ec4c22/job.log
 (1/2) xive.py:XIVE.test_intr_mode:  PASS (0.63 s)
 (2/2) xive.py:XIVE.test_storeEOI:  CANCEL: K4storeEOI feature is not Available for POWER9 / XICS (0.67 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 1
JOB HTML   : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T03.58-2ec4c22/results.html
JOB TIME   : 1.93 s

File: cpu/xive.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        if "ppc" not in os.uname()[4]:
+        self.cancel("Supported only on Power Systems")
+
+        cpu_info = genio.read_file("/proc/cpuinfo")
+        if 'POWER9' in cpu_info:
+            self.hw = "POWER9"
+        elif 'POWER10' in cpu_info:
+            self.hw = "POWER10"
+        else:
+            self.fail("Unsupported processor family")
```
Comment (nasastry) [inline cpu/xive.py:43]: Test should be cancelled here not the fail.

File: cpu/xive.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+        mode = genio.read_file("/proc/interrupts")
+        if 'XIVE' in mode:
+            self.intr = 'XIVE'
+        elif 'XICS' in mode:
+            self.intr = 'XICS'
+        else:
+            self.intr = ''
```
Comment (nasastry) [inline cpu/xive.py:51]: Assignment of null string is not required.

File: cpu/xive.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+            self.cancel("System is not Enabled with XIVE or XICS")
+
+    def test_storeEOI(self):
+        self.log.info("HW: %s Mode: %s" % (self.hw, self.intr))
+        if self.intr == 'XIVE':
+            xive_file_path = "/sys/kernel/debug/powerpc/xive"
+            if not os.path.exists(xive_file_path):
+                self.cancel("storeEOI feature is not Available %s / %s" % (self.hw, self.intr))
+            else:
+                flags = genio.read_file("/sys/kernel/debug/powerpc/xive")
```
Comment (nasastry) [inline cpu/xive.py:68]: for the sysfs filename, there is already a variable assigned `xive_file_path`. Please use that.

**Review by nasastry — CHANGES_REQUESTED** [review]

File: cpu/xive.py (hunk 1 above)
Comment (kalshett) [reply cpu/xive.py]: Changed.

File: cpu/xive.py (hunk 2 above)
Comment (kalshett) [reply cpu/xive.py]: Changed and added Cancel if mode is not xive or xics.

File: cpu/xive.py (hunk 3 above)
Comment (kalshett) [reply cpu/xive.py]: Changed.

File: cpu/xive.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        else:
+            self.cancel("Unsupported Interrupt Mode")
+
+    def test_intr_mode(self):
+        if self.intr == 'XIVE':
+            self.log.info("%s Enabled System" % self.intr)
+        elif self.intr == 'XICS':
+            self.log.info("%s Enabled System" % self.intr)
+        else:
+            self.cancel("System is not Enabled with XIVE or XICS")
```
Comment (nasastry) [inline cpu/xive.py:59]: The addition of self.cancel in setUp makes this self.cancel in function redundant. 
Do you mean self.fail here?

File: cpu/xive.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            if not os.path.exists(xive_file_path):
+                self.cancel("storeEOI feature is not Available %s / %s" % (self.hw, self.intr))
+            else:
+                flags = genio.read_file(xive_file_path)
+                match = re.search("flags=S", flags)
+                self.log.info("MATCH = %s" % match)
+                if match:
+                    self.log.info("storeEOI feature Available and 'S' flag is set for %s / %s" % (self.hw, self.intr))
+                else:
+                    self.cancel("storeEOI feature 'S' flag is not set for %s / %s" % (self.hw, self.intr))
```
Comment (nasastry) [inline cpu/xive.py:74]: In this function I am not seeing any self.fail here. 
the self.cancel here in the this function are the self.fail?
Referring to the following self.cancel 
```
self.cancel("storeEOI feature is not Available %s / %s" % (self.hw, self.intr))
self.cancel("storeEOI feature 'S' flag is not set for %s / %s" % (self.hw, self.intr))
```

**Comment by kalshett** [comment]
Reran the test on P10 default mode:
[root@ltcden14-lp10 cpu]# avocado run xive.py 
JOB ID     : f11ee5677dfdcbf8bca8a2c05cc5ac7a46b54720
JOB LOG    : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T10.26-f11ee56/job.log
 (1/2) xive.py:XIVE.test_intr_mode: PASS (0.49 s)
 (2/2) xive.py:XIVE.test_storeEOI: PASS (0.54 s)
RESULTS    : PASS 2 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /home/kal/avocado-fvt-wrapper/results/job-2021-07-20T10.26-f11ee56/results.html
JOB TIME   : 1.50 s

File: cpu/xive.py (hunk 4 above)
Comment (kalshett) [reply cpu/xive.py:59]: Here we can make it fail. Yes I can change it to fail not cancel.

**Review by nasastry — APPROVED** [review]
LGTM

**Comment by nasastry** [comment]
> This pull request **introduces 1 alert** when merging [3cfa225](https://github.com/avocado-framework-tests/avocado-misc-tests/commit/3cfa225d84f4ea6be706005b2deac5ed1747fb6e) into [ba6c4fc](https://github.com/avocado-framework-tests/avocado-misc-tests/commit/ba6c4fc2e4a6754206d90a9d0de0548cdf47f0a2) - [view on LGTM.com](https://lgtm.com/projects/g/avocado-framework-tests/avocado-misc-tests/rev/pr-52b565571c01d1374734ec75d27f7e742dd07802)
> 
> **new alerts:**
> 
> * 1 for Unused import

@kalshett Please take care of the above unused import. Thanks!! 
Other wise code looks good to me. Thanks!!

File: cpu/xive.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    XIVE test cases
+
+    :avocado: tags=cpu,power
+    '''
+
+    def setUp(self):
+        if "ppc" not in os.uname()[4]:
+            self.cancel("Supported only on Power Systems")
+
+        cpu_info = genio.read_file("/proc/cpuinfo")
```
Comment (PraveenPenguin) [inline cpu/xive.py:36]: why not we can using `sikpUnless`  class decorator 
code can be like 
```
IS_POWER10= 'Power10' in open('/proc/cpuinfo', 'r').read()
IS_POWER9 = 'POWER9' in open('/proc/cpuinfo', 'r').read()
 @skipUnless( IS_POWER10 | IS_Power9 , "Unsupported processor family")
 
 ```

File: cpu/xive.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+            self.hw = "POWER10"
+        else:
+            self.cancel("Unsupported processor family")
+
+        mode = genio.read_file("/proc/interrupts")
+        if 'XIVE' in mode:
+            self.intr = 'XIVE'
+        elif 'XICS' in mode:
+            self.intr = 'XICS'
+        else:
```
Comment (PraveenPenguin) [inline cpu/xive.py:49]: same for this check as well use as class decorator

**Review by PraveenPenguin — CHANGES_REQUESTED** [review]

File: cpu/xive.py (hunk 6 above)
Comment (kalshett) [reply cpu/xive.py:36]: This for generic power platform. So I don't think it really required to use "skipUnless", we do have other Power systems as well POWER8 for example. So I will not change this.

File: cpu/xive.py (hunk 7 above)
Comment (kalshett) [reply cpu/xive.py:49]: Maybe this we can use skipUnless.

File: cpu/xive.py (hunk 6 above)
Comment (PraveenPenguin) [reply cpu/xive.py:36]: sure make sense

Comment (PraveenPenguin) [reply cpu/xive.py:36]: may we can have skipUnless in other way 
IS_POWER8 = 'POWER8'  not in  open('/proc/cpuinfo', 'r').read()
@skipUnless(! IS_POWER8 , "Unsupported processor family")

**Review by PraveenPenguin — APPROVED** [review]

**Pull request merged**